src/ucp/core/ucp_request.inl

```diff
@@ -663,15 +663,26 @@ ucp_request_recv_data_unpack(ucp_request_t *req, const void *data,
 }
 
 static UCS_F_ALWAYS_INLINE ucp_recv_desc_t*
-ucp_recv_desc_get(ucp_worker_h worker, size_t length)
+ucp_recv_desc_alloc(ucp_worker_h worker, size_t length)
 {
-    /* TODO: Optimize */
-    uint32_t rp  = ucs_roundup_pow2(length + 1); /* make sure length is not 0 */
-    unsigned idx = ucs_popcount(worker->am_mps_map & (rp - 1));
+    size_t size = length + 1;
+    unsigned idx;
 
-    ucs_assertv(length < UINT_MAX, "length %zu", length);
+    /* - use size equal to length + 1 to avoid passing 0 to
+     *   ucs_count_leading_zero_bits()
+     * - do not roundup the size if it is pow of 2 already, to not use bigger
+     *   mpool than needed
+     */
+    idx = (sizeof(size_t) * 8) -
+           (ucs_count_leading_zero_bits(size) + !!ucs_is_pow2_or_zero(size));
```

Comment:
1. maybe ucs_count_leading_zero_bits(length - 1) + 1 ?
2. i think size can never be 0 because there is always some header?

Reply:
`ucs_count_leading_zero_bits` is undefined for zero
i got some test failure otherwise (will double check the exact test)

Reply:
Sometimes we want to keep just a data, which can be zero:
https://github.com/openucx/ucx/blob/master/src/ucp/core/ucp_am.c#L1231